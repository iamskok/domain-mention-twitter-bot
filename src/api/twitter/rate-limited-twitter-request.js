import rateLimiter from '../../services/limiter';
import logger from '../../services/logger';

const rateLimitedTwitterRequest = (
  endpoint,
  options,
  twitAuth,
  limiter = rateLimiter,
) => new Promise((resolve, reject) => {
  limiter.removeTokens(1, (error, remainingRequests) => {
    logger.log('verbose', `twitter/rateLimitedTwitterRequest remaining requests: ${remainingRequests}`);

    if (error) {
      logger.log('error', '`twitter/rateLimitedTwitterRequest` limiter', {
        errorObject: error,
      });

      reject(error);
    } else {
      twitAuth.get(endpoint, options, async (twitError, data, response) => {
        if (twitError) {
          logger.log('error', '`twitter/rateLimitedTwitterRequest` received error from `twitAuth.get`', {
            errorObject: twitError.message,
          });

          return reject(twitError);
        }

        if (response.statusCode !== 200) {
          logger.log('error', '`twitter/rateLimitedTwitterRequest` `twitAuth.get` received non 200 response status code', {
            errorObject: error.message,
          });

          return reject(twitError);
        }

        return resolve(data);
      });
    }
  });
});

export { rateLimitedTwitterRequest };
