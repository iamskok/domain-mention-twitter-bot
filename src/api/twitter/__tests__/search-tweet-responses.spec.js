import * as rateLimitedTwitterRequestModule from '../rate-limited-twitter-request';
import searchTweetResponses from '../search-tweet-responses';
import tweetsMock from '../../../mocks/tweets';

const rateLimitedTwitterRequestMock = (endpoint, options, twitAuth, limiter) => {
  const isReply = !!options.q.includes('to:');

  if (isReply) {
    return new Promise((resolve) => {
      resolve({
        statuses: tweetsMock.filter(({ in_reply_to_status_id_str: replyId }) => !!replyId),
      });
    });
  }

  return new Promise((resolve) => {
    resolve({
      statuses: tweetsMock.filter(({ quoted_status_id_str: quoteId }) => !!quoteId),
    });
  });
};

describe('searchTweetResponses', () => {
  it('returns single tweet reply with child quote and reply', () => {
    jest.spyOn(rateLimitedTwitterRequestModule, 'rateLimitedTwitterRequest')
      .mockImplementation(rateLimitedTwitterRequestMock);

    expect(searchTweetResponses(tweetsMock[0], 'reply', [])).resolves.toMatchSnapshot();
  });

  it('returns single tweet quote with child quote and reply', () => {
    jest.spyOn(rateLimitedTwitterRequestModule, 'rateLimitedTwitterRequest')
      .mockImplementation(rateLimitedTwitterRequestMock);

    expect(searchTweetResponses(tweetsMock[0], 'quote', [])).resolves.toMatchSnapshot();
  });

  it('returns no quotes', () => {
    jest.spyOn(rateLimitedTwitterRequestModule, 'rateLimitedTwitterRequest')
      .mockImplementation(rateLimitedTwitterRequestMock);

    expect(searchTweetResponses(tweetsMock[7], 'quote', [])).resolves.toEqual([]);
  });

  it('returns no replies', () => {
    jest.spyOn(rateLimitedTwitterRequestModule, 'rateLimitedTwitterRequest')
      .mockImplementation(rateLimitedTwitterRequestMock);

    expect(searchTweetResponses(tweetsMock[7], 'quote', [])).resolves.toEqual([]);
  });
});
