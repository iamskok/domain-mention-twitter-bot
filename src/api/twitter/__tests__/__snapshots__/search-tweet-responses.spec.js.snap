// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`searchTweetResponses returns no quotes 1`] = `
Array [
  Object {
    "created_at": "Wed Oct 10 01:30:24 +0000 2020",
    "entities": undefined,
    "id_str": "111111111111111111",
    "in_reply_to_status_id_str": null,
    "quoted_status_id_str": "000000000000000000",
    "quotes": Array [
      Object {
        "created_at": "Wed Oct 10 01:50:24 +0000 2020",
        "entities": undefined,
        "id_str": "333333333333333333",
        "in_reply_to_status_id_str": null,
        "quoted_status_id_str": "111111111111111111",
        "quotes": Array [],
        "replies": Array [],
        "text": "Yo!",
        "user": Object {
          "name": undefined,
          "profile_image_url_https": undefined,
          "screen_name": "Lisa",
        },
      },
    ],
    "replies": Array [
      Object {
        "created_at": "Wed Oct 10 01:40:24 +0000 2020",
        "entities": undefined,
        "id_str": "444444444444444444",
        "in_reply_to_status_id_str": "111111111111111111",
        "quoted_status_id_str": null,
        "quotes": Array [],
        "replies": Array [],
        "text": "How are you guys?",
        "user": Object {
          "name": undefined,
          "profile_image_url_https": undefined,
          "screen_name": "Beth",
        },
      },
    ],
    "text": "Hey!!",
    "user": Object {
      "name": undefined,
      "profile_image_url_https": undefined,
      "screen_name": "John",
    },
  },
]
`;

exports[`searchTweetResponses returns single tweet quote with child quote and reply 1`] = `
Array [
  Object {
    "created_at": "Wed Oct 10 01:34:24 +0000 2020",
    "entities": undefined,
    "id_str": "222222222222222222",
    "in_reply_to_status_id_str": "000000000000000000",
    "quoted_status_id_str": null,
    "quotes": Array [
      Object {
        "created_at": "Wed Oct 10 02:40:24 +0000 2020",
        "entities": undefined,
        "id_str": "666666666666666666",
        "in_reply_to_status_id_str": null,
        "quoted_status_id_str": "222222222222222222",
        "quotes": Array [],
        "replies": Array [],
        "text": "Hello friends!",
        "user": Object {
          "name": undefined,
          "profile_image_url_https": undefined,
          "screen_name": "Cindy",
        },
      },
    ],
    "replies": Array [
      Object {
        "created_at": "Wed Oct 10 02:30:24 +0000 2020",
        "entities": undefined,
        "id_str": "555555555555555555",
        "in_reply_to_status_id_str": "222222222222222222",
        "quoted_status_id_str": null,
        "quotes": Array [],
        "replies": Array [],
        "text": "How are you guys?",
        "user": Object {
          "name": undefined,
          "profile_image_url_https": undefined,
          "screen_name": "Bob",
        },
      },
    ],
    "text": "Hello world",
    "user": Object {
      "name": undefined,
      "profile_image_url_https": undefined,
      "screen_name": "Marry",
    },
  },
]
`;
